### PR TITLE
[MLX] Add Q5_K quantization support

### LINKED ISSUE #20614: [MLX][Good first issue] Add Q5K quantization support
## Summary

Add native **fused Q5_K Metal kernels** for the MLX backend, matching the existing
**Q6_K** ([#20004]) and **Q4_K** ([#20172]) support. Q5_K is currently **not
handled at all** — it is rejected on the export side and has no MLX lowering.
We want Q5_K linear/embedding to read the raw GGUF `block_q5_K` directly in fused
custom Metal kernels, the same way Q6_K and Q4_K do, with no export-time repack.

## Background

`ExportableGGUFTensor` (`extension/llm/export/gguf.py`) lowers a quantized
linear/embedding to a `torchao::dequantize_gguf -> linear/embedding` subgraph.
The MLX pattern handlers in
`backends/mlx/custom_kernel_ops/gguf/patterns.py` match that subgraph and lower
it to fused Metal kernels without materializing the dequantized weight.

Today only two K-quant formats are wired end to end, and both make good
references because Q5_K borrows from each:

- **Q6_K** → fused custom Metal kernels in
  `backends/mlx/custom_kernel_ops/gguf/q6k/`. A `block_q6_K` struct plus dequant
  helpers live in `q6k/common.py` (`_Q6K_HEADER`), and `q6k/linear.py` emits two
  kernels ported from llama.cpp:
  - `M == 1` (decode): a fused mat-vec kernel (`kernel_mul_mv_q6_K_f32_impl`).
  - `M > 1` (prefill): a tiled simdgroup mat-mat kernel (`kernel_mul_mm`).
  - dynamic `M`: both are emitted and selected at runtime via an `IfNode`.
  Q6_K is the reference for the **high-bit (`qh`) packing** — Q5_K has the same
  notion of an extra bit per quant stored in a separate `qh` array.

- **Q4_K** → fused custom Metal kernels in
  `backends/mlx/custom_kernel_ops/gguf/q4k/` (`_Q4K_HEADER` in `q4k/common.py`).
  Q4_K is the reference for the **affine super-block layout**: a super-block
  scale `d` and min `dmin`, with 6-bit packed sub-block scales/mins unpacked via
  `get_scale_min_k4`. Q5_K shares this exact scale/min layout.

Q5_K = Q4_K's affine `d`/`dmin` + 6-bit packed scales/mins, **plus** a Q6_K-style
high-bit array. Each weight is a 5-bit code: the low 4 bits from `qs` and the 5th
bit from `qh`.

## Task

Implement fused Q5_K Metal kernels (linear + embedding) so that Q5_K consumes the
raw `block_q5_K` directly, and register Q5_K on the export side so it lowers
through the MLX pattern handlers.

Concretely:

1. **`extension/llm/export/gguf.py`** — register Q5_K so `dequantize_gguf` /
   `ExportableGGUFTensor` accept it:
   - add `GGML_Q5_K = 13` (the `gguf.GGMLQuantizationType` id for Q5_K),
   - add `_Q5_K_BLOCK_BYTES = 2 + 2 + 12 + QK_K // 8 + QK_K // 2  # 176`,
   - add `"q5_k"` to `_GGML_ID_BY_TYPE` and `_BLOCK_BYTES_BY_TYPE`.
   (The `gguf` python package can already dequantize Q5_K, so the fake/dequant
   path works once the type is registered. A torchao `Int*Tensor` conversion
   path like `_q4_k_fields` / `_q6_k_fields` is **not** required — the MLX kernels
   read the raw blob.)

2. **`backends/mlx/custom_kernel_ops/gguf/q5k/common.py`** (new) — add a
   `_Q5K_HEADER` Metal header with the `block_q5_K` struct and dequant helpers
   (per-element for embedding, vectorized `half4x4` for matmul), plus `QK_K` /
   `K_SCALE_SIZE` / `Q5K_BLOCK_BYTES` constants. Port from llama.cpp
   `dequantize_q5_K` / `dequantize_row_q5_K` (`ggml-common.h` /
   `ggml-metal.metal`). Reuse `get_scale_min_k4` / `get_scale_min_k4_just2`
   exactly as in `q4k/common.py`, and add the 5th bit from `qh` the way Q6_K
   reads its high bits:

   ```c
   #define QK_K 256
   #define K_SCALE_SIZE 12
   typedef struct {
       half     d;                    // super-block scale for the quantized scales
       half     dmin;                 // super-block scale for the quantized mins
       uint8_t  scales[K_SCALE_SIZE]; // 6-bit packed scales + mins (same as Q4_K)
       uint8_t  qh[QK_K/8];           // quants, high bit (32 bytes)
       uint8_t  qs[QK_K/2];           // quants, low 4 bits (128 bytes)
   } block_q5_K;                      // 176 bytes
   ```

   Dequant for a 5-bit code `q` in sub-block `s`:
   `q = (qs_nibble) | (((qh >> bit) & 1) << 4)`, value `= d * sc[s] * q - dmin * m[s]`
   (affine, same shape as Q4_K but with `q` in `0..31`).

3. **`backends/mlx/custom_kernel_ops/gguf/q5k/linear.py`** (new) — provide
   `emit_linear` (and an `eager_linear` reference) with mat-vec (decode),
   mat-mat (prefill), and dynamic-`M` `IfNode` emission, mirroring
   `q6k/linear.py` (port llama.cpp `kernel_mul_mv_q5_K_f32_impl` and the Q5_K
   `kernel_mul_mm` variant).

4. **`backends/mlx/custom_kernel_ops/gguf/q5k/embedding.py`** (new) — provide
   `emit_embedding`, a per-element Q5_K dequant gather, mirroring
   `q6k/embedding.py`. Add `q5k/__init__.py`.

5. **`backends/mlx/custom_kernel_ops/gguf/patterns.py`** — wire Q5_K in:
   - add `"q5_k"` to `_LINEAR_TYPES` and `_EMBEDDING_TYPES`,
   - add a `self.ggml_type == "q5_k"` dispatch branch in both
     `GGUFQuantizedLinearHandler.__call__` and
     `GGUFQuantizedEmbeddingHandler.__call__` (import `emit_linear` /
     `emit_embedding` from `.q5k`),
   - update the module docstring to list Q5_K.

## Testing

Add Q5_K coverage mirroring the existing Q4_K / Q6_K cases in
`backends/mlx/custom_kernel_ops/gguf/test/test_linear.py` and
`test_embedding.py`:

- add a `make_q5_k_blob(N, K, seed)` fixture next to `make_q6_k_blob` /
  `make_q4_k_blob`, and register it in `_BLOB_MAKERS`,
- add `ggml_type="q5_k"` configs to `GGUFLinearTest.get_test_configs`,
  `GGUFLinearDynamicTest.get_test_configs`, and the embedding test configs,
- use the gguf-exact reference `weight.dequantize(torch.float32)` for Q5_K
  (`_fp32_linear_reference`), same as Q6_K — Q5_K has no repack path, so no
  special-casing like Q4_K's `_q4k_mlx_native_dequant` is needed.

Run on an Apple-silicon machine:

```
python -m executorch.backends.mlx.custom_kernel_ops.gguf.test.test_linear run -v
python -m executorch.backends.mlx.custom_kernel_ops.gguf.test.test_embedding run -v
```

## Pointers

- Reference (high bits / kernel structure): `backends/mlx/custom_kernel_ops/gguf/q6k/{common,linear,embedding}.py`
- Reference (affine `d`/`dmin`, `get_scale_min_k4`): `backends/mlx/custom_kernel_ops/gguf/q4k/common.py`
- Pattern handlers: `backends/mlx/custom_kernel_ops/gguf/patterns.py`
- Export-side type registration: `extension/llm/export/gguf.py`
- Tests: `backends/mlx/custom_kernel_ops/gguf/test/test_linear.py`, `test_embedding.py`
- Reference PRs: #20004 (Q6_K), #20172 (Q4_K)

## Attribution

The Q5_K block layout and Metal dequant helpers should be ported from llama.cpp
(`ggml-common.h` / `ggml-metal.metal`: `block_q5_K`, `dequantize_q5_K`,
`kernel_mul_mv_q5_K_f32_impl`, `kernel_mul_mm`), which is MIT-licensed
(Copyright (c) 2023-2024 The ggml authors). Keep inline `ported from ...` notes
as in the Q6_K / Q4_K kernels.

### PR DESCRIPTION
**Pull request by JaynouOliver**
## Summary

Adds fused **Q5_K** Metal kernels (linear + embedding) for the MLX backend and registers Q5_K on the export side, matching the existing **Q6_K** (#20004) and **Q4_K** (#20172) support. Resolves #20614.

Q5_K = Q4_K's affine super-block (`d`/`dmin` + 6-bit packed sub-block scales/mins via `get_scale_min_k4`) **plus** a Q6_K-style high-bit array: each weight is a 5-bit code whose low 4 bits come from `qs` and whose 5th bit comes from `qh`. The kernels consume the raw `block_q5_K` directly — no export-time repack.

## Changes

- **`extension/llm/export/gguf.py`** — register `GGML_Q5_K = 13`, `_Q5_K_BLOCK_BYTES = 176` (`2 + 2 + 12 + QK_K//8 + QK_K//2`), add `"q5_k"` to `_GGML_ID_BY_TYPE` / `_BLOCK_BYTES_BY_TYPE`. (No torchao `Int*Tensor` conversion path — the MLX kernels read the raw blob.)
- **`backends/mlx/custom_kernel_ops/gguf/q5k/`** (new):
  - `common.py` — `_Q5K_HEADER` with the `block_q5_K` struct, `dequant_q5k_elem` (per-element, embedding) and `dequantize_q5_K_16` (vectorized `half4x4`, matmul), reusing `get_scale_min_k4` / `get_scale_min_k4_just2`.
  - `linear.py` — `emit_linear`: mat-vec (decode), tiled simdgroup mat-mat (prefill), dynamic-`M` `IfNode`. Byte-wise decode with the 5th bit from `qh`; affine `d*scale` / `dmin*min` per sub-block.
  - `embedding.py` — `emit_embedding`: per-element Q5_K dequant gather.
  - `__init__.py`.
- **`backends/mlx/custom_kernel_ops/gguf/patterns.py`** — add `"q5_k"` to `_LINEAR_TYPES` / `_EMBEDDING_TYPES` and a dispatch branch in both handlers.
- **Tests** — `make_q5_k_blob` + `q5_k` configs in `test/test_linear.py` and `test/test_embedding.py`; Q5_K coverage in `extension/llm/export/test/test_gguf.py`.

Kernels ported from llama.cpp (`ggml-common.h` / `ggml-metal.metal`: `block_q5_K`, `dequantize_q5_K`, `kernel_mul_mv_q5_K_f32_impl`, `kernel_mul_mm`), MIT-licensed (Copyright (c) 2023-2024 The ggml authors); inline `ported from ...` notes kept as in the Q6_K / Q4_K kernels.

## Validation

Done locally on Apple silicon (macOS 26.4, MLX 0.31.2):

- ✅ **Export side** — `extension/llm/export/test/test_gguf.py` passes (9/9), including the new Q5_K cases (`dequantize` == `gguf.dequantize` exactly; `torchao::dequantize_gguf` op; `torch.export` lowering; unsupported-type guard).
- ✅ **Kernel math** — the per-element (`dequant_q5k_elem`), vectorized (`dequantize_q5_K_16`), and mat-vec arithmetic were each transcribed and checked numerically against `gguf.dequantize` / `x @ Wᵀ`: exact for the dequant paths, ~3e-7 rel for the mat-vec (fp32 rounding).
- ✅ MLX backend imports with Q5_K wired into both pattern handlers.
- ⚠️ **On-device Metal kernel execution** (`test_linear.py` / `test_embedding.py run`) was **not** run locally: building the `op_test_runner` compiles MLX from source via `xcrun metal`, which needs full Xcode (only Command Line Tools available on the dev box). **Relying on CI to run these on the GPU.** The mat-mat kernel is structurally identical to the merged Q4_K/Q6_K kernels and differs only in the (numerically validated) `dequantize_q5_K_16` weight decode.

## Test plan

On an Apple-silicon machine:

```
python -m executorch.backends.mlx.custom_kernel_ops.gguf.test.test_linear run -v
python -m executorch.backends.mlx.custom_kernel_ops.gguf.test.test_embedding run -v
```

### CODE REVIEW

**Comment by JaynouOliver** [comment]
hi @metascroy , can you have a look please?

**Comment by metascroy** [comment]
@claude review this code

**Comment by metascroy** [comment]
Thanks @JaynouOliver!

Overall it looks good, but a few comments:

* Dequant (dequantize_q5_K_16) needs no change (exact port of llama.cpp dequantize_q5_K)
* Matmat is a faithful port of llama.cpp: same 64×32 tiles, NK=32, tile indexing, K-loop, simdgroup ops, and matching launch params (128 threads / 4 simdgroups, grid = ceil(M/32) × ceil(N/64)). So overall it looks good. But **you should shrink sb[4096] to sb[1024]** (only 1024 halves are used). [Note sa[4096] is correctly sized (reused as the 8 KB float output-staging buffer)].

Also parallelize the write out: see https://github.com/pytorch/executorch/pull/20643 for similar improvements on q4k/q6k.

* Matvec
Numerically equivalent to llama.cpp, but not a port (different thread to work mapping, get_scale_min_k4 instead of inline kmask, folded accumulation). It also uses N_R0=2, nsg=2 (borrowed from Q4_K) vs llama.cpp Q5_K's N_R0=1, nsg=2. **Replace the Q4_K-derived matvec with a faithful port of llama.cpp's kernel_mul_mv_q5_K_f32_impl.**

**Comment by JaynouOliver** [comment]
thank you will get this done shortly

**Comment by JaynouOliver** [comment]
Thanks @metascroy for the review! Addressed all three:
**Dequant (`dequantize_q5_K_16`)** — left unchanged (exact port of llama.cpp `dequantize_q5_K`), as you noted.

**Mat-mat**
- Shrank `sb[4096]` -> `sb[1024]` (only 16 `ib` slots x 64 halves are used; `sa` stays `[4096]` as the 8 KB float output-staging buffer).
- Parallelized the write-out across all 128 threads, mirroring the Q4_K/Q6_K change in #20643.

**Mat-vec** — replaced the Q4_K-derived kernel with a faithful port of llama.cpp `kernel_mul_mv_q5_K_f32_impl` (`N_R0 = 1`, `NSG = 2`): `tid = tiisg/4`, `ix = tiisg%4` thread map, inline kmask scale/min unpacking, and the nibble sum / 5th-bit indicator sum (`acc1`/`acc2`) accumulated separately then folded with `d*scale` / `dmin*min`.

Validation (no Metal locally — only Command Line Tools on the dev box, so the `metal` compiler for `op_test_runner` isn't available; relying on CI for the on-device run):
- The faithful mat-vec arithmetic was checked lane-by-lane against `x @ Wᵀ` (rel ~2e-7 vs the gguf reference).
- All Q5_K configs (decode mat-vec, prefill mat-mat, dynamic-M IfNode, embedding) export to a valid `.pte` through the MLX partitioner.

**Comment by metascroy** [comment]
@claude review this code

**Comment by metascroy** [comment]
Looks great @JaynouOliver!

There was a small compile error (using half as a variable name, which is a dtype in metal). I pushed a fix, and if tests pass, I'll merge.

Thanks for the contribution!

**Review by metascroy — APPROVED** [review]
